### PR TITLE
Navigator: Leverage overlapping fields in logic to save RAM by makein…

### PR DESCRIPTION
**Pull request by LorenzMeier**
…g them overlap in memory as well.

This change does save quite a bit of RAM when handling mission items in main memory. @bkueng @julianoes @davids5 FYI. This also needs a bit of review in its usage, but will then bring down the RAM consumed by dataman by a few K.

TODO:

- [x] Bitfield for bool flags
- [x] Remove origin field

### CODE REVIEW

File: src/modules/navigator/navigation.h [hunk 1]
```diff
@@ -100,26 +100,34 @@ enum ORIGIN {
...
+	int16_t do_jump_mission_index;		/**< index where the do jump will go to                 */
+	uint16_t do_jump_repeat_count;		/**< how many times do jump needs to be done            */
+	uint16_t do_jump_current_count;		/**< count how many times the jump has been done	*/
+	uint16_t origin;					/**< where the waypoint has been generated		*/
+	bool loiter_exit_xtrack;			/**< exit xtrack location: 0 for center of loiter wp, 1 for exit location */
+	bool force_heading;					/**< heading needs to be reached ***/
+	bool altitude_is_relative;			/**< true if altitude is relative from start point	*/
+	bool autocontinue;					/**< true if next waypoint should follow after this one */
+	int8_t frame;						/**< mission frame ***/
+	int8_t loiter_direction;			/**< 1: positive / clockwise, -1, negative.		*/
```
Comment (dagar) [inline src/modules/navigator/navigation.h]: Combine loiter_radius and direction like in mavlink?

File: src/modules/navigator/navigation.h [hunk 2]
```diff
@@ -100,26 +100,34 @@ enum ORIGIN {
...
+			float ___lon_float;			/**< padding */
+			float altitude;				/**< altitude in meters	(AMSL)			*/
+		};
+		float params[7];				/**< array to store mission command values for MAV_FRAME_MISSION ***/
+	};
+	enum NAV_CMD nav_cmd;				/**< navigation command					*/
+	int16_t do_jump_mission_index;		/**< index where the do jump will go to                 */
+	uint16_t do_jump_repeat_count;		/**< how many times do jump needs to be done            */
+	uint16_t do_jump_current_count;		/**< count how many times the jump has been done	*/
+	uint16_t origin;					/**< where the waypoint has been generated		*/
```
Comment (dagar) [inline src/modules/navigator/navigation.h]: I don't think origin is actually used.

**Comment by LorenzMeier** [comment]
I think these are all safe / reasonable changes. This should save about 50% of the previous RAM and support 50 waypoints in about 5K of RAM. @bkueng we still might need to optimize depending on how much the new logger pulls on resource constrained systems.

**Comment by LorenzMeier** [comment]
jenkins test this please.

**Comment by davids5** [comment]
This change is also binary incompatible with existing files. Can re add back the structure packing and detect the size not modulo sizeof(union ) == 0 and wipe it?

**Comment by LorenzMeier** [comment]
@davids5 Yep, need to do that. I believe dataman already does it, but need to re-check.

**Comment by LorenzMeier** [comment]
Jenkins test this please.

**Comment by LorenzMeier** [comment]
With the last commits its now fool proof and ready to go in after further SITL testing.

**Comment by davids5** [comment]
@LorenzMeier 

Short of using a tool like UAVCAN uses to crc32 the type data. We could use the poor mans not so automatic and prone to human error version.

this is what I am thinking of and rough pseudo code that should scale long term.

Add a in dataman.h

```
#define DATAMAN_MK_REVISION(_version, struct_name)  {.version=_version, .size =sizeof(struct_name)}
```
add in each header (or in the case of an auto generates ones find a good place for it)
```
#define mission_item_version DATAMAN_MK_REVISION(1, mission_item)
```
in dataman.c
```
#typdef  uint8_t dmr_sizet ; // lord help us if this has to be more than 8 bits!

typedef struct dataman_revison_item_t {
dmr_sizet version;
dmr_sizet size;
} dataman_revison_item_t ;

#define countoftypes  4
```
then define a const

```
struct dataman_revison {
uint8_t  size_of; 
uint8_t  nTypes;
 dataman_revison_item_t  version v[N] 
dmr_sizet version v[countoftypes] 
} = {
.size_of = sizeof(dmr_sizet),
.nTypes = <countoftypes>
.v={{mission_item_version },
{mission_version },
{fence_vertex_version }},
};
```

Thoughs?

**Pull request merged**